common/s3util.cpp

```diff
@@ -493,4 +507,124 @@ shared_ptr<S3Util> S3Util::BuildS3Util(
       new S3Util(bucket, aws_config, options, read_ratelimit_mb, write_ratelimit_mb));
 }
 
+
+S3Concurrent::S3Concurrent(const string &bucket, const int upload_MBps)
+  : bucket_(bucket),
+    upload_MBps_(upload_MBps),
+    executor_(2),  // n_threads
+    transfer_config_(&executor_) {
+
+  const uint64_t MiB = 1024*1024;
+  const uint64_t max_heap_size = 400 * MiB;
+  const uint64_t buffer_size = 25 * MiB;
+  int max_connections = 16; // upper limit, good for 10Gbps.
+
+  // throttle concurrency according to desired rate.
+  // 2 threads allows for some file interleaving.
+  // approx 70 MB/sec speed per connection.
+  if (upload_MBps_ != 0) {
+    max_connections = std::min(max_connections, std::max(1, (int)(upload_MBps_/70)));
+  }
```

Comment:
will throttle by `max_connections` more safe or throttle by actual network speed, regardless to avoid overloading host.

Reply:
Docs recommend using max_connections to manage bandwidth usage:
https://awscli.amazonaws.com/v2/documentation/api/latest/topic/s3-config.html#max-bandwidth
Using AWS DefaultRateLimiter as well as targeting the concurrency would allow for a bit of safety while avoiding the problems with too many blocked threads.
I'll add it in and test.

Reply:
Ratelimiter added.